File: muzei-nationalgeographic/src/main/java/de/msal/muzei/nationalgeographic/ItemDeserializer.java
package de.msal.muzei.nationalgeographic;

import android.text.Html;
import com.google.gson.*;
import de.msal.muzei.nationalgeographic.model.Item;

import java.lang.reflect.Type;

public class ItemDeserializer implements JsonDeserializer<Item> {

   @Override
   public Item deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) {
      JsonObject jsonObject = json.getAsJsonObject();
      JsonElement image = jsonObject.get("image");

      Item item;
      if (image != null) { // old/private api
         item = context.deserialize(image, Item.class);
         item.setInternal(jsonObject.get("internal").getAsBoolean());
         item.setPageUrlPhotoOfTheDay(jsonObject.get("pageUrl").getAsString());
         item.setPublishDate(jsonObject.get("publishDate").getAsString());
         return item;
      } else { // new/public api
         item = new Gson().fromJson(jsonObject, Item.class);
      }

      // clean up photographer info
      item.setPhotographer(stripHtml(item.getPhotographer())
            .replace("Photograph by ", "")
            .replace(", National Geographic Your Shot", "")
            .replace(", National Geographic", "")
            .replace(", Your Shot", "")
            .replace(", My Shot", "")
            .trim()
      );
      return item;
   }

   private String stripHtml(String html) {
      if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
         return Html.fromHtml(html, Html.FROM_HTML_MODE_LEGACY).toString();
      } else {
         //noinspection deprecation
         return Html.fromHtml(html).toString().trim();
      }
   }

}

File: muzei-nationalgeographic/src/main/kotlin/de/msal/muzei/nationalgeographic/ItemDeserializer.kt
package de.msal.muzei.nationalgeographic

import com.github.sisyphsu.dateparser.DateParserUtils
import com.google.gson.*
import com.google.gson.JsonParseException
import de.msal.muzei.nationalgeographic.model.Image
import de.msal.muzei.nationalgeographic.model.Item
import java.lang.reflect.Type

/**
 * "Flattens" the Json as the relevant data is nested and also cleans up some fields
 */
class ItemDeserializer : JsonDeserializer<Item> {

   @Throws(JsonParseException::class)
   override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): Item {
      val item = GsonBuilder()
            .registerTypeAdapter(Image::class.java, ImageDeserializer())
            .create()
            .fromJson(json.asJsonObject["entity"], Item::class.java)

      // clean up photographer info
      item.credit = item.credit
            ?.stripHtml()
            ?.extractName()
      item.contributor = item.contributor
            ?.stripHtml()
            ?.extractName()
      // clean up description
      item.caption = item.caption
            ?.stripHtml()

      // add date
      item.entityLabel?.replace("pod-", "")?.let {
         item.date = DateParserUtils.parseCalendar(it)
      }

      return item
   }

   private fun String?.stripHtml(): String? {
      // don't need full html escaping. just remove the <p> tags as nothing else seems to exist
      return this?.replace("<p>", "")?.replace("</p>", "")?.trim()
      //return when {
      //   this != null -> Html.escapeHtml(this)
      //   else -> this
      //}
   }

   private fun String?.extractName(): String? {
      return this
            ?.replace("Photograph by ", "", true)
            ?.replace(", National Geographic Your Shot", "", true)
            ?.replace(", National Geographic Image Collection", "", true)
            ?.replace(", National Geographic", "", true)
            ?.replace(", Your Shot", "", true)
            ?.replace(", My Shot", "", true)
            ?.replace(", nat geo image collection", "", true)
            ?.trim()
   }

}
